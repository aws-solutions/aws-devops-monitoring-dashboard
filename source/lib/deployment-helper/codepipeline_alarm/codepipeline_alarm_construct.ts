// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as cw from 'aws-cdk-lib/aws-cloudwatch';
import * as logs from 'aws-cdk-lib/aws-logs';
import * as events from 'aws-cdk-lib/aws-events';
import * as targets from 'aws-cdk-lib/aws-events-targets';
import { addCfnSuppressRules } from '@aws-solutions-constructs/core';

export interface PipelineMetricProps {
  readonly pipelineName: string;
  readonly paramCreateLogGroup: cdk.CfnParameter;
  readonly logGroupName: string;
  readonly metricNameSpace: string;
  readonly alarmName: string;
}

export class PipelineAlarmConstruct extends Construct {
  constructor(scope: Construct, id: string, props: PipelineMetricProps) {
    super(scope, id);

    /**
     * Create CloudWatch log group if YES is selected for creating new log group
     */
    const pipelineLogGroup = new logs.LogGroup(this, 'PipelineLogGroup', {
      logGroupName: props.logGroupName,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      retention: logs.RetentionDays.THREE_MONTHS
    });

    const condCreateLogGroupTrue = new cdk.CfnCondition(this, 'CreateLogGroupTrue', {
      expression: cdk.Fn.conditionEquals('YES', props.paramCreateLogGroup)
    });

    const pipelineLogGroupCfnRef = pipelineLogGroup.node.defaultChild as logs.CfnLogGroup;
    pipelineLogGroupCfnRef.cfnOptions.condition = condCreateLogGroupTrue;

    addCfnSuppressRules(pipelineLogGroupCfnRef, [
      {
        id: 'W84',
        reason: 'CloudWatch LogGroups are encrypted by default.'
      }
    ]);

    // Get iLogGroup reference - works for new or existing log group
    const pipelineIlogGroup: logs.ILogGroup = logs.LogGroup.fromLogGroupName(
      this,
      'PipelineILogGroup',
      props.logGroupName
    );

    /**
     * Create CloudWatch Events Rule for AWS CodePipeline execution state change
     */
    const pipelineEventRulePattern = {
      source: ['aws.codepipeline'],
      'detail-type': ['CodePipeline Pipeline Execution State Change'],
      detail: {
        state: ['FAILED', 'SUCCEEDED']
      }
    };

    const pipelineEventsRule = new events.Rule(this, 'PipelineEventsRule', {
      eventPattern: pipelineEventRulePattern
    });

    pipelineEventsRule.addTarget(new targets.CloudWatchLogGroup(pipelineIlogGroup));

    /**
     * Create CloudWatch Metric Filter for AWS CodePipeline failed executions
     */
    const metricName = `FailedExecution-${props.pipelineName}`;
    const metricFilter = new logs.MetricFilter(this, 'FailedPipelineExecutionMetricFilter', {
      filterPattern: {
        logPatternString: '{($.detail.state = "FAILED")}'
      },
      logGroup: pipelineIlogGroup,
      metricName: metricName,
      metricNamespace: props.metricNameSpace,
      defaultValue: 0,
      metricValue: '1'
    });

    const metricFilter_cfn_ref = metricFilter.node.defaultChild as logs.CfnMetricFilter;
    metricFilter_cfn_ref.addDependsOn(pipelineLogGroupCfnRef);

    /**
     * Create CloudWatch alarm for AWS CodePipeline failed execution metric
     */
    const alarm = new cw.Alarm(this, 'AppAlarm', {
      alarmName: props.alarmName,
      metric: new cw.Metric({
        namespace: props.metricNameSpace,
        metricName: metricName,
        period: cdk.Duration.minutes(1),
        statistic: 'Sum'
      }),
      threshold: 1,
      comparisonOperator: cw.ComparisonOperator.GREATER_THAN_OR_EQUAL_TO_THRESHOLD,
      evaluationPeriods: 1,
      datapointsToAlarm: 1,
      treatMissingData: cw.TreatMissingData.IGNORE,
      alarmDescription: 'Goes to alarm state when there is one failed code pipeline execution on the most recent check.'
    });

    const alarmResource = alarm.node.findChild('Resource') as cw.CfnAlarm;
    addCfnSuppressRules(alarmResource, [
      {
        id: 'W28',
        reason: 'Static names chosen intentionally to provide fixed name structure required in the solution'
      }
    ]);
  }
}
