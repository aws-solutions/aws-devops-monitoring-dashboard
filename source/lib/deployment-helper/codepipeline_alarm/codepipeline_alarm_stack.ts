// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

/**
 * @description
 * Creates a Canary that monitors an http or https url, and an alarm that will
 * send a signal to the solution each time the app is not 100% available.
 * Optionally, an S3 bucket for canary artifacts and access logging bucket are
 * created.
 * @author @aws-solutions
 */

import * as cdk from 'aws-cdk-lib';
import { Construct, IConstruct } from 'constructs';
import { PipelineAlarmConstruct } from './codepipeline_alarm_construct';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { addCfnSuppressRules } from '@aws-solutions-constructs/core';
import * as iam from 'aws-cdk-lib/aws-iam';
import { NagSuppressions } from 'cdk-nag';

export interface PipelineAlarmProps extends cdk.StackProps {
  readonly solutionId: string;
  readonly solutionVersion: string;
}

export class PipelineAlarmStack extends cdk.Stack {
  /**
   * @function Object() { [native code] }
   * @param {Construct} scope - parent of the construct
   * @param {string} id - identifier for the object
   * @param props - properties for the construct
   */
  constructor(scope: Construct, id: string, props: PipelineAlarmProps) {
    super(scope, id, props);

    //=========================================================================
    // PARAMETERS
    //=========================================================================

    /**
     * @description name of the CodePipeline
     * @type {cdk.CfnParameter}
     */
    const paramCodePipelineName = new cdk.CfnParameter(this, 'CodePipelineName', {
      description: 'Name of the CodePipeline. Change the default value as needed.',
      type: 'String'
    });

    /**
     * @description Create a new log group? YES or NO. A new log group is created during initial stack deployment.
     * For subsequent stack updates, recommend to select NO to reuse the existing log group.
     * @type {cdk.CfnParameter}
     */
    const paramCreateLogGroup = new cdk.CfnParameter(this, 'CreateLogGroup', {
      description:
        'Create a new log group? Select YES to create a new log group if this is the first time you deploy the stack. For subsequent stack update, recommend to select NO to reuse the existing log group. Specify the Log group name below.',
      type: 'String',
      default: 'YES',
      allowedValues: ['YES', 'NO']
    });

    /**
     * @description Name of the CodePipeline log group to which the pipeline events are sent
     * @type {cdk.CfnParameter}
     */
    const paramLogGroupName = new cdk.CfnParameter(this, 'LogGroupName', {
      description:
        'Name of the CodePipeline log group to which the pipeline events are sent. Change the default value as needed.',
      type: 'String',
      default: 'my-codepipeline-log-group'
    });

    /**
     * @description name of the source repository for the code pipeline
     * monitored
     * @type {cdk.CfnParameter}
     */
    const paramRepoName = new cdk.CfnParameter(this, 'RepoName', {
      description:
        'Name of source repository for the code pipeline. It will be part of the alarm name. Change the default value as needed.',
      type: 'String'
    });

    const parameterMetaData = {
      'AWS::CloudFormation::Interface': {
        ParameterGroups: [
          {
            Label: {
              default: 'Events Rule Configuration'
            },
            Parameters: ['CodePipelineName', 'CreateLogGroup', 'LogGroupName']
          },
          {
            Label: {
              default: 'Alarm Configuration'
            },
            Parameters: ['RepoName']
          }
        ],
        ParameterLabels: {
          CodePipelineName: {
            default: 'CodePipeline Name'
          },
          CreateLogGroup: {
            default: 'Create a new log Group?'
          },
          LogGroupName: {
            default: 'Log Group Name'
          },
          RepoName: {
            default: 'Repository Name'
          }
        }
      }
    };
    this.templateOptions.metadata = parameterMetaData;

    const alarmName =
      props.solutionId + '-[' + paramCodePipelineName.valueAsString + ']-[' + paramRepoName.valueAsString + ']-MTTR';

    const metricNameSpaceMapping = new cdk.CfnMapping(this, 'CodePipelineMetrics', {
      mapping: {
        CodePipelineMetrics: {
          NameSpace: `CodePipeline/${props.solutionId}/Pipelines`
        }
      }
    });

    new PipelineAlarmConstruct(this, 'CodePipelineAlarm', {
      pipelineName: paramCodePipelineName.valueAsString,
      paramCreateLogGroup: paramCreateLogGroup,
      logGroupName: paramLogGroupName.valueAsString,
      metricNameSpace: metricNameSpaceMapping.findInMap('CodePipelineMetrics', 'NameSpace'),
      alarmName: alarmName
    });

    class AddCfnSuppressRules implements cdk.IAspect {
      public visit(node: IConstruct): void {
        if (node instanceof lambda.CfnFunction) {
          addCfnSuppressRules(node, [
            {
              id: 'W58',
              reason: 'Lambda has the required permission to write CloudWatch Logs through a custom policy.'
            },
            {
              id: 'W89',
              reason: 'No need to deploy this Lambda to a VPC.'
            },
            {
              id: 'W92',
              reason: 'No need for simultaneous executions.'
            }
          ]);
        } else if (node instanceof iam.CfnPolicy) {
          addCfnSuppressRules(node, [
            {
              id: 'W12',
              reason:
                'Resource * is required by the Lambda Execution role, so that the Lambda can add ResourcePolicies to all required resources.'
            }
          ]);
        }
      }
    }

    cdk.Aspects.of(this).add(new AddCfnSuppressRules());

    // Add cdk-nag suppression
    NagSuppressions.addStackSuppressions(this, [
      {
        id: 'AwsSolutions-IAM5',
        reason:
          'Resource * is required by the Lambda Execution role, so that the Lambda can add ResourcePolicies to all required resources.'
      },
      {
        id: 'AwsSolutions-IAM4',
        reason: 'The managed policy is automatically generated by CDK to be used by lambda service role.'
      },
      {
        id: 'AwsSolutions-L1',
        reason: 'The lambda function and its runtime are automatically generated by CDK.'
      }
    ]);

    //=========================================================================
    // OUTPUTS
    //=========================================================================
    new cdk.CfnOutput(this, 'SolutionVersion', {
      value: props.solutionVersion,
      description: 'Version for DevOps Monitoring Dashboard on AWS solution'
    });

    new cdk.CfnOutput(this, 'Log Group Name', {
      value: paramLogGroupName.valueAsString,
      description: 'Name of the log group'
    });

    new cdk.CfnOutput(this, 'Alarm Name', {
      value: alarmName,
      description: 'Name of the CodePipeline alarm'
    });
  }
}
