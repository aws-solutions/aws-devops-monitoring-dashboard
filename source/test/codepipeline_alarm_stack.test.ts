// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import * as cdk from 'aws-cdk-lib';
import { Stack } from 'aws-cdk-lib';
import { PipelineAlarmStack } from '../lib/deployment-helper/codepipeline_alarm/codepipeline_alarm_stack';
import { expect as expectCDK, haveResourceLike, SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';

const SOLUTION_ID = 'SO0143';
const DIST_VERSION = 'v1.0.0';

/*
 * Snapshot test for codepipeline alarm stack.
 */
test('Snapshot test for codepipeline alarm stack', () => {
  const app = new cdk.App();
  const stack = new PipelineAlarmStack(app, 'CodePipelineAlarm', {
    solutionId: SOLUTION_ID,
    solutionVersion: DIST_VERSION
  });
  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});

describe('PipelineAlarmStack', () => {
  let stack: Stack;

  beforeAll(() => {
    const app = new cdk.App();

    stack = new PipelineAlarmStack(app, 'CodePipelineAlarm', {
      solutionId: SOLUTION_ID,
      solutionVersion: DIST_VERSION
    });
  });

  const logGroupNameRef = {
    Ref: 'LogGroupName'
  };

  test('it has a CloudWatch Alarm that checks for pipeline failures', () => {
    expectCDK(stack).to(
      haveResourceLike('AWS::CloudWatch::Alarm', {
        ComparisonOperator: 'GreaterThanOrEqualToThreshold',
        Threshold: 1
      })
    );
  });

  test('it has a LogGroup for CodePipeline alarms', () => {
    expectCDK(stack).to(
      haveResourceLike('AWS::Logs::LogGroup', {
        LogGroupName: logGroupNameRef,
        RetentionInDays: 90
      })
    );
  });

  test('it has an EventsRule for CodePipeline execution state changes', () => {
    expectCDK(stack).to(
      haveResourceLike('AWS::Events::Rule', {
        State: 'ENABLED'
      })
    );
  });

  test('it has a MetricFilter checking for pipeline failures in the log group', () => {
    expectCDK(stack).to(
      haveResourceLike('AWS::Logs::MetricFilter', {
        LogGroupName: logGroupNameRef,
        FilterPattern: '{($.detail.state = "FAILED")}'
      })
    );
  });
});
