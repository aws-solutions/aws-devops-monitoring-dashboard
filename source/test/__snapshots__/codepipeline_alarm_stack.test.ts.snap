// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test for codepipeline alarm stack 1`] = `
{
  "Conditions": {
    "CodePipelineAlarmCreateLogGroupTrue2C38BECE": {
      "Fn::Equals": [
        "YES",
        {
          "Ref": "CreateLogGroup",
        },
      ],
    },
  },
  "Mappings": {
    "CodePipelineMetrics": {
      "CodePipelineMetrics": {
        "NameSpace": "CodePipeline/SO0143/Pipelines",
      },
    },
  },
  "Metadata": {
    "AWS::CloudFormation::Interface": {
      "ParameterGroups": [
        {
          "Label": {
            "default": "Events Rule Configuration",
          },
          "Parameters": [
            "CodePipelineName",
            "CreateLogGroup",
            "LogGroupName",
          ],
        },
        {
          "Label": {
            "default": "Alarm Configuration",
          },
          "Parameters": [
            "RepoName",
          ],
        },
      ],
      "ParameterLabels": {
        "CodePipelineName": {
          "default": "CodePipeline Name",
        },
        "CreateLogGroup": {
          "default": "Create a new log Group?",
        },
        "LogGroupName": {
          "default": "Log Group Name",
        },
        "RepoName": {
          "default": "Repository Name",
        },
      },
    },
    "cdk_nag": {
      "rules_to_suppress": [
        {
          "id": "AwsSolutions-IAM5",
          "reason": "Resource * is required by the Lambda Execution role, so that the Lambda can add ResourcePolicies to all required resources.",
        },
        {
          "id": "AwsSolutions-IAM4",
          "reason": "The managed policy is automatically generated by CDK to be used by lambda service role.",
        },
        {
          "id": "AwsSolutions-L1",
          "reason": "The lambda function and its runtime are automatically generated by CDK.",
        },
      ],
    },
  },
  "Outputs": {
    "AlarmName": {
      "Description": "Name of the CodePipeline alarm",
      "Value": {
        "Fn::Join": [
          "",
          [
            "SO0143-[",
            {
              "Ref": "CodePipelineName",
            },
            "]-[",
            {
              "Ref": "RepoName",
            },
            "]-MTTR",
          ],
        ],
      },
    },
    "LogGroupName": {
      "Description": "Name of the log group",
      "Value": {
        "Ref": "LogGroupName",
      },
    },
    "SolutionVersion": {
      "Description": "Version for DevOps Monitoring Dashboard on AWS solution",
      "Value": "v1.0.0",
    },
  },
  "Parameters": {
    "CodePipelineName": {
      "Description": "Name of the CodePipeline. Change the default value as needed.",
      "Type": "String",
    },
    "CreateLogGroup": {
      "AllowedValues": [
        "YES",
        "NO",
      ],
      "Default": "YES",
      "Description": "Create a new log group? Select YES to create a new log group if this is the first time you deploy the stack. For subsequent stack update, recommend to select NO to reuse the existing log group. Specify the Log group name below.",
      "Type": "String",
    },
    "LogGroupName": {
      "Default": "my-codepipeline-log-group",
      "Description": "Name of the CodePipeline log group to which the pipeline events are sent. Change the default value as needed.",
      "Type": "String",
    },
    "RepoName": {
      "Description": "Name of source repository for the code pipeline. It will be part of the alarm name. Change the default value as needed.",
      "Type": "String",
    },
  },
  "Resources": {
    "AWS679f53fac002430cb0da5b7982bd22872D164C4C": {
      "DependsOn": [
        "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
      ],
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W58",
              "reason": "Lambda has the required permission to write CloudWatch Logs through a custom policy.",
            },
            {
              "id": "W89",
              "reason": "No need to deploy this Lambda to a VPC.",
            },
            {
              "id": "W92",
              "reason": "No need for simultaneous executions.",
            },
          ],
        },
      },
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "e845402ce43b66fc6f20df4a239f20f8662eb6c7f920b94cf6542dd0e64ce0f7.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CodePipelineAlarmAppAlarmC62B9C08": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W28",
              "reason": "Static names chosen intentionally to provide fixed name structure required in the solution",
            },
          ],
        },
      },
      "Properties": {
        "AlarmDescription": "Goes to alarm state when there is one failed code pipeline execution on the most recent check.",
        "AlarmName": {
          "Fn::Join": [
            "",
            [
              "SO0143-[",
              {
                "Ref": "CodePipelineName",
              },
              "]-[",
              {
                "Ref": "RepoName",
              },
              "]-MTTR",
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "DatapointsToAlarm": 1,
        "EvaluationPeriods": 1,
        "MetricName": {
          "Fn::Join": [
            "",
            [
              "FailedExecution-",
              {
                "Ref": "CodePipelineName",
              },
            ],
          ],
        },
        "Namespace": {
          "Fn::FindInMap": [
            "CodePipelineMetrics",
            "CodePipelineMetrics",
            "NameSpace",
          ],
        },
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 1,
        "TreatMissingData": "ignore",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "CodePipelineAlarmFailedPipelineExecutionMetricFilter36FFD4B2": {
      "DependsOn": [
        "CodePipelineAlarmPipelineLogGroupB061BB52",
      ],
      "Properties": {
        "FilterPattern": "{($.detail.state = "FAILED")}",
        "LogGroupName": {
          "Ref": "LogGroupName",
        },
        "MetricTransformations": [
          {
            "DefaultValue": 0,
            "MetricName": {
              "Fn::Join": [
                "",
                [
                  "FailedExecution-",
                  {
                    "Ref": "CodePipelineName",
                  },
                ],
              ],
            },
            "MetricNamespace": {
              "Fn::FindInMap": [
                "CodePipelineMetrics",
                "CodePipelineMetrics",
                "NameSpace",
              ],
            },
            "MetricValue": "1",
          },
        ],
      },
      "Type": "AWS::Logs::MetricFilter",
    },
    "CodePipelineAlarmPipelineEventsRule97CA1513": {
      "Properties": {
        "EventPattern": {
          "detail": {
            "state": [
              "FAILED",
              "SUCCEEDED",
            ],
          },
          "detail-type": [
            "CodePipeline Pipeline Execution State Change",
          ],
          "source": [
            "aws.codepipeline",
          ],
        },
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":logs:",
                  {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":log-group:",
                  {
                    "Ref": "LogGroupName",
                  },
                ],
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "CodePipelineAlarmPipelineLogGroupB061BB52": {
      "Condition": "CodePipelineAlarmCreateLogGroupTrue2C38BECE",
      "DeletionPolicy": "Delete",
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W84",
              "reason": "CloudWatch LogGroups are encrypted by default.",
            },
          ],
        },
      },
      "Properties": {
        "LogGroupName": {
          "Ref": "LogGroupName",
        },
        "RetentionInDays": 90,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "EventsLogGroupPolicyCodePipelineAlarmPipelineEventsRule4FBE0864CustomResourcePolicyD35BFE30": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W12",
              "reason": "Resource * is required by the Lambda Execution role, so that the Lambda can add ResourcePolicies to all required resources.",
            },
          ],
        },
      },
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "logs:PutResourcePolicy",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "logs:DeleteResourcePolicy",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "EventsLogGroupPolicyCodePipelineAlarmPipelineEventsRule4FBE0864CustomResourcePolicyD35BFE30",
        "Roles": [
          {
            "Ref": "AWS679f53fac002430cb0da5b7982bd2287ServiceRoleC1EA0FF2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "EventsLogGroupPolicyCodePipelineAlarmPipelineEventsRule4FBE0864F2A10BA1": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "EventsLogGroupPolicyCodePipelineAlarmPipelineEventsRule4FBE0864CustomResourcePolicyD35BFE30",
      ],
      "Properties": {
        "Create": {
          "Fn::Join": [
            "",
            [
              "{"service":"CloudWatchLogs","action":"putResourcePolicy","parameters":{"policyName":"CodePipelineAlarmEventsLogGroupPolicyCodePipelineAlarmPipelineEventsRule4FBE0864184E82F4","policyDocument":"{\\"Statement\\":[{\\"Action\\":[\\"logs:PutLogEvents\\",\\"logs:CreateLogStream\\"],\\"Effect\\":\\"Allow\\",\\"Principal\\":{\\"Service\\":\\"events.amazonaws.com\\"},\\"Resource\\":\\"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":logs:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":log-group:",
              {
                "Ref": "LogGroupName",
              },
              ":*\\"}],\\"Version\\":\\"2012-10-17\\"}"},"physicalResourceId":{"id":"EventsLogGroupPolicyCodePipelineAlarmPipelineEventsRule4FBE0864"}}",
            ],
          ],
        },
        "Delete": "{"service":"CloudWatchLogs","action":"deleteResourcePolicy","parameters":{"policyName":"CodePipelineAlarmEventsLogGroupPolicyCodePipelineAlarmPipelineEventsRule4FBE0864184E82F4"},"ignoreErrorCodesMatching":"400"}",
        "InstallLatestAwsSdk": true,
        "ServiceToken": {
          "Fn::GetAtt": [
            "AWS679f53fac002430cb0da5b7982bd22872D164C4C",
            "Arn",
          ],
        },
        "Update": {
          "Fn::Join": [
            "",
            [
              "{"service":"CloudWatchLogs","action":"putResourcePolicy","parameters":{"policyName":"CodePipelineAlarmEventsLogGroupPolicyCodePipelineAlarmPipelineEventsRule4FBE0864184E82F4","policyDocument":"{\\"Statement\\":[{\\"Action\\":[\\"logs:PutLogEvents\\",\\"logs:CreateLogStream\\"],\\"Effect\\":\\"Allow\\",\\"Principal\\":{\\"Service\\":\\"events.amazonaws.com\\"},\\"Resource\\":\\"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":logs:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":log-group:",
              {
                "Ref": "LogGroupName",
              },
              ":*\\"}],\\"Version\\":\\"2012-10-17\\"}"},"physicalResourceId":{"id":"EventsLogGroupPolicyCodePipelineAlarmPipelineEventsRule4FBE0864"}}",
            ],
          ],
        },
      },
      "Type": "Custom::CloudwatchLogResourcePolicy",
      "UpdateReplacePolicy": "Delete",
    },
  },
}
`;
